docs/reference/mapping/fields/synthetic-source.asciidoc

```diff
@@ -0,0 +1,170 @@
+* Synthetic `_source` can be used with indices that contain only these field
+types:
+
+** <<boolean-synthetic-source,`boolean`>>
+** <<numeric-synthetic-source,`byte`>>
+** <<numeric-synthetic-source,`double`>>
+** <<numeric-synthetic-source,`float`>>
+** <<geo-point-synthetic-source,`geo_point`>>
+** <<numeric-synthetic-source,`half_float`>>
+** <<numeric-synthetic-source,`integer`>>
+** <<ip-synthetic-source,`ip`>>
+** <<keyword-synthetic-source,`keyword`>>
+** <<numeric-synthetic-source,`long`>>
+** <<numeric-synthetic-source,`scaled_float`>>
+** <<numeric-synthetic-source,`short`>>
+** <<text-synthetic-source,`text`>>
+
+[[synthetic-source-modifications]]
+===== Synthetic source modifications
+
+When synthetic `_source` is enabled, retrieved documents undergo some
+modifications compared to the original JSON.
+
+[[synthetic-source-modifications-alphabetical]]
+====== Alphabetical sorting
+Synthetic `_source` fields are sorted alphabetically. For example:
+
+[source,console,id=synthetic-source-sorted-example]
+----
+PUT idx/_doc/1
+{
+  "foo": 1,
+  "bar": 2,
+  "baz": 3
+}
+----
+// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]
+
+Will become:
+
+[source,console-result]
+----
+{
+  "bar": 2,
+  "baz": 3,
+  "foo": 1
+}
+----
+// TEST[s/^/{"_source":/ s/\n$/}/]
+
+[[synthetic-source-modifications-leaf-arrays]]
+====== Arrays moved to leaf fields
+Synthetic `_source` arrays are moved to leaves. For example:
+
+[source,console,id=synthetic-source-leaf-arrays-example]
+----
+PUT idx/_doc/1
+{
+  "foo": [
+    {
+      "bar": 1
+    },
+    {
+      "bar": 2
+    }
+  ]
+}
+----
+// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]
+
+Will become:
+
+[source,console-result]
+----
+{
+  "foo": {
+    "bar": [1, 2]
+  }
+}
+----
+// TEST[s/^/{"_source":/ s/\n$/}/]
+
+[[synthetic-source-modifications-field-names]]
+====== Fields named as they are mapped
+Synthetic source names fields as they are named in the mapping. When used
+with <<dynamic,dynamic mapping>>, fields with dots (`.`) in their names are, by
+default, interpreted as multiple objects. For example:
+
+[source,console,id=synthetic-source-objecty-example]
+----
+PUT idx/_doc/1
+{
+  "foo.bar.baz": 1
+}
+----
+// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]
+
+Will become:
+
+[source,console-result]
+----
+{
+  "foo": {
+    "bar": {
+      "baz": 1
+    }
+  }
+}
+----
+// TEST[s/^/{"_source":/ s/\n$/}/]
+
+To change this behavior and store names without splitting them on dots, set
+<<subobjects>> to `false`. For example:```
+
+[source,console,id=synthetic-dot-example]
+----
+PUT idx
+{
+  "mappings": {
+    "subobjects": false,
+    "_source": {
+      "synthetic": true
+    }
+  }
+}
+
+PUT idx/_doc/1
+{
+  "foo.bar.baz": 1
+}
+----
+// TEST[s/^/DELETE idx\n/]
+// TEST[s/$/\nGET idx\/_doc\/1?filter_path=_source\n/]
+
+Will remain as:
+
+[source,console-result]
+----
+{
+  "foo.bar.baz": 1
+}
+----
```

Comment:
I wonder what is specific of synthetic source about how fields are mapped: isn't this all about dynamic mappings? Did you mean to say that synthetic _source recreate the objects structure or not depending on how fields are mapped?

Reply:
It recreates the structure precisely as the objects are mapped. At first I tried to explicitly create a field mapped as `foo.bar.baz` but the mapping infrastructure unraveled it so I went with this. I'll make an example, one moment.

Reply:
Ok, I guess I was wondering why this needs to be specifically explained, isn't it the expected behaviour? :)

Reply:
I think you are explaining how dynamic mappings work in combination with subobjects fails, and that synthetic source respects that, but you could omit this here, because it would be surprising and would need to be mentioned if it worked otherwise.

Reply:
I suppose. But it was kind of surprising to me that it *did* work with subobjects. Maybe that's just stuck in my head.

Reply:
Scratch that, maybe this is not so obvious :)
One thing to mention could be that you always recreate the object structure despite dots in fields names were provided in the first place. Why not always flatten by the way? After all, there are two ways to provide that document and that leads to the same mapping, you have 50% chances to pick the variant that was sent in the first place.

By the way we may start accepting docs with subobjects even when subobjects are disabled and treat them like dotted fields, but I think the current behaviour of "flattening" would still be good in that case.

Reply:
I've removed the section

Reply:
Oh no! I removed the section!

> Why not always flatten by the way?

We asked some kibana friends and they said folks liked things shaped this way instead of flattened. The objects feel right to folks. Also, I figured if we were guessing anyway the "more objecty" approach was probably more likely to match what users sent. I figured flattened fields with dots in them is more rare.

Reply:
fine with me. I think you removed only the section with subobjects false and you could make the remaining section about recreating objects shorter by saying that you follow the mappings structure hence prefer nested objects over dots in fields names but the opposite when subobjects are disabled.

Reply:
> I think you removed only the section with subobjects false and you could make the remaining section about recreating objects shorter by saying that you follow the mappings structure hence prefer nested objects over dots in fields names but the opposite when subobjects are disabled.

Two things:
1. Don't use the word `nested` unless you mean it. I get scared.
2. Do you want me to change the words above? Could you suggest a change? I think the "dynamic mappings work like BLAH" bit is useful because it explains where those objects come from. I don't mean to explain all of dynamic mappings, just enough to give folks a hint.